src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,263 @@
+ * index to the array [15, 21, 31, 41, 51, 61], where each element denotes how
+ * many function evaluations take place in each subinterval. The GSL
+ * documentation mentions that a higher-order rule serves better for smooth
+ * functions, whereas a lower-order rule saves time for functions with local
+ * difficulties, such as discontinuities.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::StandardGaussKronrod>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, const double lower_boundary,
+                    const double upper_boundary, const double tolerance_abs,
+                    const int key, const double tolerance_rel = 0.) const {
+    double result;
+    double error;
+    detail::disable_gsl_error_handling();
+    const int status_code = gsl_integration_qag(
+        gsl_integrand(std::forward<IntegrandType>(integrand)), lower_boundary,
+        upper_boundary, tolerance_abs, tolerance_rel, this->max_intervals_, key,
+        this->workspace_.get(), &result, &error);
+    detail::check_status_code(status_code);
+    return result;
+  }
+};
+
+/*!
+ * The algorithm for "IntegrableSingularitiesPresent" concentrates new,
+ * increasingly smaller subintervals around an unknown singularity and makes
+ * successive approximations to the integral which should converge towards a
+ * limit. The integration region is defined by `lower_boundary` and
+ * `upper_boundary`.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::IntegrableSingularitiesPresent>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, const double lower_boundary,
+                    const double upper_boundary, const double tolerance_abs,
+                    const double tolerance_rel = 0.) const noexcept {
+    double result;
+    double error;
+    detail::disable_gsl_error_handling();
+    const int status_code = gsl_integration_qags(
+        gsl_integrand(std::forward<IntegrandType>(integrand)), lower_boundary,
+        upper_boundary, tolerance_abs, tolerance_rel, this->max_intervals_,
+        this->workspace_.get(), &result, &error);
+    detail::check_status_code(status_code);
+    return result;
+  }
+};
+
+/*!
+ * The algorithm for "IntegrableSingularitiesKnown" uses user-defined
+ * subintervals given by a vector of doubles `points`, where each element
+ * denotes an interval boundary.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::IntegrableSingularitiesKnown>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, std::vector<double> points,
```

Comment:
why not take `const std::vector<double>?` you don't move below

Reply:
The function changes the std::vector<double>, I think. Hence it takes a pointer to points. When I set it to const, I get a compiler error that says the argument would lose the const qualifier.

Reply:
It's not clear to me why the points should changed. Do you understand what is trying to change them and why?

Reply:
The `points` describe the boundaries of each subinterval. I figured, that when the algorithm increases the number of subintervals to reach the desired accuracy it writes the new boundaries to `points`.

Reply:
Okay, I see now... This is actually a bug in GSL because (looking at the GSL source code) `points` should be `const` because you are telling the integrator where the singularities are. It would be rather disastorous (especially in the current case) if it tried to change the size of the member because `std::vector` allocates with `new`, while `gsl` allocates with `malloc`, and these two are incompatible. I think (unfortunately) the best solution here is actually:
```cpp
 double operator()(IntegrandType&& integrand, const std::vector<double>& points,
                  const double tolerance_abs,
                  const double tolerance_rel = 0.) const noexcept {
  double result;
  double error;
  detail::disable_gsl_error_handling();
  // The const_cast is necessary because GSL has a weird interface where
  // the number of singularities does not change, but it doesn't take
  // the argument as a `const double*`. However, the first thing
  // `gsl_integration_qagp` does internally is forward to another function that
  // does take `points` by `const double*`. If `gsl_integration_qagp` were to
  // change the size of `points` this code would be severely broken because
  // `std::vector` allocates with `new`, while GSL would likely use `malloc` (or
  // some other C allocator). Mixing (de)allocators in such a way is undefined
  // behavior.
  const int status_code = gsl_integration_qagp(
      gsl_integrand(std::forward<IntegrandType>(integrand)),
      const_cast<double*>(points.data()), // NOLINT
      points.size(), tolerance_abs, tolerance_rel, this->max_intervals_,
      this->workspace_.get(), &result, &error);
  detail::check_status_code(status_code);
  return result;
}
```
This reduces memory allocations and makes

Reply:
Thank you a lot!

Reply:
You're welcome :)